src/MockFactory.ts

```diff
@@ -64,111 +65,146 @@ const executeOperation = (
-    this.schema = generateMockSchema(schema)
-    this.mocks = mocks || {}
+  private maybeAddTypenameToDocument = (
+    document: DocumentNode,
+    overrideAddTypename: Boolean | undefined,
+  ) => {
+    const addTypename = overrideAddTypename ?? this.addTypename
+    // ? parse(print(addTypenameToDocument(document)))
+    return addTypename ? addTypenameToDocument(document) : document
+  }
+
+  constructor(
+    schema: IntrospectionQuery,
+    { mocks, addTypename = false }: Options = {},
+  ) {
+    this.schema = buildMockSchema(schema)
+    this.mocks = mocks ?? {}
+    this.addTypename = addTypename
   }
 
-  mockFragment = <TData = any>(
+  mockFragment = <TData = any, TVariables = any>(
     fragment: DocumentNode,
-    { mocks }: Options = { mocks: {} },
+    options: MockOptions<TVariables> = {},
   ): TData => {
-    const mainDefinition = getMainDefinition(fragment)
-    const fragmentWithTypename = addTypenameToDocument(fragment)
+    const { variables = {}, mocks = {}, addTypename } = options
+
+    const fragmentDocument = this.maybeAddTypenameToDocument(
+      fragment,
+      addTypename,
+    )
+    const mainDefinition = getMainDefinition(fragmentDocument)
 
     if (mainDefinition.kind !== 'FragmentDefinition') {
       throw new Error(
         'MockFactory: mockFragment only accepts fragment documents',
       )
     }
 
-    /* eslint-disable graphql/template-strings */
     const typeName = mainDefinition.typeCondition.name.value
     const fieldName = `mock__${typeName}`
-    const query = gql`
+    const query = parse(/* GraphQL */ `
       query {
         ${fieldName} {
-          ...${mainDefinition.name?.value}
+          ...${mainDefinition.name.value}
         }
       }
-      ${fragmentWithTypename}
-    `
-    /* eslint-enable graphql/template-strings */
+      ${print(fragmentDocument)}
+    `)
 
     const mockedSchema = this.mockSchema({ mocks })
 
-    const result = executeOperation(mockedSchema, query)
+    const result = executeOperation(mockedSchema, query, variables)
 
     if (!result.data || result.data[fieldName] === undefined) {
       throw new Error(
-        `Unable to generate mock data for ${
-          mainDefinition.name.value || 'fragment'
-        }. This could be a result of missing mock resolvers or an incorrect fragment structure.`,
+        [
+          `Unable to generate mock data for ${
+            mainDefinition.name.value || 'fragment'
+          }. This could be a result of missing mock resolvers or an incorrect fragment structure.`,
+          print(fragment),
+        ].join('\n'),
       )
     }
-    return { ...result.data[fieldName] }
+
+    return JSON.parse(JSON.stringify(result.data[fieldName])) as TData
   }
 
   mockQuery = <TData = any, TVariables = any>(
     query: DocumentNode,
-    options: { variables?: TVariables; mocks?: ResolverMap } = {},
+    options: {
+      variables?: TVariables
+      mocks?: ResolverMap
+      addTypename?: true | false
```

Comment:
Does the boolean type not work here?

Reply:
Typescript was having issues when I used `Boolean` for some reason. I think I was able to work around it